broadinstitute/single_cell_portal_core
# Fix "Annotated scatter" in new Explore UX (flagged, SCP-5155)

This fixes handling of numeric annotations in the new, flagged Explore UX.  "Annotated scatter" plots now show as expected.

Previously, in #1782, the prototype UX worked for group annotations, but broke for numeric annotations.  While that severity is high, the affected numeric annotations currently only account for 5% of default annotations among the first 20 public studies in https://singlecell.broadinstitute.org, so the incidence of this problem would be low.  Fortunately the issue was caught before enabling the related feature flag in an A/B test, so this affected no users.

### Video
Here's how it looks.

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/a8094245-6f6c-4b5b-9568-1fee1e1a4792

### Test
The bug emerged in a hard-to-test corner of SCP React code.  I refactored to ease testing, and added regression tests.

To manually verify these changes:
* Go to a study that contains numeric annotations.  If on staging, the [release engineer tests](https://docs.google.com/document/d/1rKcOQPtkbhZCbTqxTnaT-4E29VLzkJrm0kAoWjLx3QA/edit#bookmark=id.8k715bvl2ls) include one.
* Ensure `show_explore_tab_ux_updates` feature flag is enabled
* Confirm "Annotated scatter" appears as a disabled tab in 0-gene view
* Search a gene
* Confirm "Annotated scatter" appears as the active, enabled tab in 1-gene view
* Search another gene, for 2 total
* Confirm "Annotated scatter" appears as a disabled tab in 2-gene view
* Search another gene, for 3 total
* Confirm "Annotated scatter" appears as a disabled tab in >= 3-gene view

This satisfies SCP-5155.
